scripts_python/create_hdf5_dataset.py

```diff
@@ -0,0 +1,255 @@
+
+
+def _add_all_subjs_to_database(args, chosen_subjs: List[str],
+                               groups_config: Dict, hdf5_dir, hdf5_filename,
+                               dwi_ml_dir):
+    """
+    Generate a dataset from a group of dMRI subjects with multiple bundles.
+    All bundles are merged as a single whole-brain dataset in voxel space.
+    If wished, all intermediate steps are saved on disk in the hdf5 folder.
+    """
+    # Cast space from str to Space
+    if args.space == 'Space.Vox':
+        space = Space.Vox
+    elif args.space == 'Space.RASMM':
+        space = Space.RASMM
+    else:
+        space = Space.VOXMM
+
+    # Add data to database
+    hdf5_file_path = hdf5_dir.joinpath("{}.hdf5".format(hdf5_filename))
+    with h5py.File(hdf5_file_path, 'w') as hdf_file:
+        # Save version and configuration
+        hdf_file.attrs['version'] = 1
+        hdf_file.attrs['chosen_subjs'] = chosen_subjs
+        hdf_file.attrs['groups'] = str(groups_config)
+        if args.step_size is not None:
+            hdf_file.attrs['step_size'] = args.step_size
+        else:
+            hdf_file.attrs['step_size'] = 'Not defined by user'
+        hdf_file.attrs['bundles'] = [b for b in args.bundles]
+        hdf_file.attrs['space'] = args.space
+
+        # Add data one subject at the time
+        logging.info("Processing {} subjects : "
+                     "{}".format(len(chosen_subjs), chosen_subjs))
+        for subj_id in chosen_subjs:
+            logging.info("\n *Processing subject: {}".format(subj_id))
+            subj_input_dir = dwi_ml_dir.joinpath(subj_id)
+
+            # Add subject to hdf database
+            subj_hdf = hdf_file.create_group(subj_id)
+
+            # Prepare subject folder for intermediate files
+            subj_intermediate_path = hdf5_dir.joinpath(subj_id)
+            if args.save_intermediate:
+                subj_intermediate_path.mkdir()
+
+            # Find subject's mask
+            if args.mask:
+                subj_mask_file = subj_input_dir.joinpath(args.mask)
+                subj_mask_data = nib.load(subj_mask_file)
+                subj_mask_data = subj_mask_data.get_fdata().astype(np.bool)
+
+            # Add the subj data based on groups in the json config file
+            # (inputs and others. All nifti)
+            for group in groups_config:
+                logging.debug('Processing group {}...'.format(group))
+                file_list = groups_config[group]
+                group_data, group_affine, group_header = process_group(
+                    group, file_list, args.save_intermediate, subj_input_dir,
+                    subj_intermediate_path, subj_mask_data)
+                logging.debug('...done. Now creating dataset from group.')
+                subj_hdf.create_dataset(group, data=group_data)
+                logging.debug('...done.')
+
+            # Reparing header. Should be arranged in Dipy 1.2.
+            group_header['srow_x'] = group_affine[0, :]
```

Comment:
Why is this needed? I have never seen this requirement...

Reply:
Hmmm.. I honestly don't remember, it's been too long. I think I was just printing everything as I was coding and I think I saw that it contained only zeros, or something. I surely didn't invent that it would be fixed in 1.2. Maybe I discussed it with François?

Do I keep it like that?

Reply:
I have no idea, I'll try to investigate a bit this week or next week

Reply:
Ok, so I checked quickly, and Dipy is not involved in fetching the header or affine (it is fetched straight from `nibabel.load().header`). This means that this "repair" is useful only if the header is already broken, in which case, it should be fixed elsewhere. 

Also, do we actually need the whole header? Some data might live in the same voxel space, but in a different world space (e.g. T1 vs. DWI), which means the header/affine might be a different, no?

Reply:
Since the advent of the StatefulTractogram, is this still needed ? Header properties of nifitis and tractograms are weird and often misused and I wouldn't play with them too much unless strictly necessary.

Reply:
I will try to test this this week. I was already using stateful tractogram when I wrote the code. I will have to check why I had problems.

Reply:
> Also, do we actually need the whole header? Some data might live in the same voxel space, but in a different world space (e.g. T1 vs. DWI), which means the header/affine might be a different, no?

I checked with tractoflow output, and the affine from T1 and dwi are the same. Don't they always need to have the same if they are registered? I'm not sure.

However for the header, the dwi has:
```
'srow_x': array([-1. ,  0. ,  0. , 77.5], dtype=float32),
 'srow_y': array([   0. ,    1. ,    0. , -113.5], dtype=float32),
 'srow_z': array([  0.  ,   0.  ,   1.  , -63.25], dtype=float32),
```
Which is the same value as the affine, but the T1 has
```
 'srow_x': array([0., 0., 0., 0.], dtype=float32),
 'srow_y': array([0., 0., 0., 0.], dtype=float32),
 'srow_z': array([0., 0., 0., 0.], dtype=float32),
```
Which explains that I had issues. The step before registration is Crop_T1, and then the header fits with the affine. So possible solutions: 1) we don't care that headers don't fit, as long as affines fit. 2) We don't care about neither. 3) We keep this trick to "repair" the header.

What is your opinion?